### PR TITLE
Control package

### PR DESCRIPTION
**Pull request by jnn16012**
# Description

This package includes control systems for the rudder and sail.

## Type of change

- [ ] Bug fix
- [ ] New feature
- [x] New ROS package

# How Has This Been Tested?

Testing can be done by: 
* running the rostest.test file, e.g. with the commando "rostest ctrl_pkg rostest.test"
* running the libtest_rudder.py,  e.g. with the commando "rosrun ctrl_pkg libtest_rudder.py"
* running the libtest_sail.py,  e.g. with the commando "rosrun ctrl_pkg libtest_sail.py" 

- [x] Unit test: Provide instructions on how to run
- [ ] Integration test: Provide instructions on how to run
- [x] ROS node test: Provide instructions on how to run
- [ ] System test: Provide instructions on how to run

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added docstring to all my functionality that follows the style guidelines of this project
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by AutoSail-MDH** [comment]
Please fill out the form for future documentation

File: ctrl_pkg/config/config.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+max_roll: 30 #0.52359877559 # pi/6
+
+# --- Rudder Configs ---
+rudder_limits: 45 #0.7853981633974483  # pi/4 radians
+pid_coefficients:
+    kp: 1
+    ki: 0.1
+    kd: 0.05
+rudder_upper_threshold: 1.5
+rudder_lower_threshold: 0.5
```
Comment (AutoSail-MDH) [inline ctrl_pkg/config/config.yaml]: Add a new line

File: ctrl_pkg/launch/node.launch [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<launch>
+   <node pkg="ctrl_pkg" name="sail_controller" type="sail_node_pid.py">
+      <rosparam command="load" file="$(find ctrl_pkg)/config/config.yaml"/>
+   </node>
+   <node pkg="ctrl_pkg" name="rudder_controller" type="rudder_node.py">
+      <rosparam command="load" file="$(find ctrl_pkg)/config/config.yaml"/>
+   </node>
+</launch>
```
Comment (AutoSail-MDH) [inline ctrl_pkg/launch/node.launch]: Add a new line

File: ctrl_pkg/scripts/sail_node_old.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        rospy.loginfo("Sail angle: %f", new_sail_angle_rad)
+        rospy.loginfo("Trim degree: %f", trim_degree)
+
+        # Publish the sail angle
+        sail_angle.publish(new_sail_angle_rad)
+        sail_servo.publish(trim_degree)
+
+        # Keep sync with the ROS frequency
+        rate.sleep()
+
```
Comment (AutoSail-MDH) [inline ctrl_pkg/scripts/sail_node_old.py]: Is this file even used?

File: ctrl_pkg/setup.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env python
+from setuptools import setup
+from catkin_pkg.python_setup import generate_distutils_setup
+
+setup_args = generate_distutils_setup(
+    packages=['ctrl'],
+    package_dir={'': 'src'},
+    install_requires=[
```
Comment (AutoSail-MDH) [inline ctrl_pkg/setup.py]: Remove install_requires since it will be handled in the package instead

File: ctrl_pkg/src/ctrl/pid.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        self.last_time = _now
+
+        return output
+
+    @property
+    def setpoint(self):
+        return self._setpoint
+
+    @setpoint.setter
+    def setpoint(self, new_setpoint):
```
Comment (AutoSail-MDH) [inline ctrl_pkg/src/ctrl/pid.py:89]: Missing docstrings

File: ctrl_pkg/tests/integrationtest_ruddercontroller.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        rot_quat = rot.as_quat()
+        navheading_msg.orientation(rot_quat)
+        self.navheading_pub.publish(navheading_msg)
+
+        velocity_msg = TwistWithCovarianceStamped
+        velocity_msg.twist.twist.linear.x = 1
+        velocity_msg.twist.twist.linear.y = 0
+        self.velocity_pub.publish(velocity_msg)
+
+        self.assertEqual(rudder_angle, 1)
```
Comment (AutoSail-MDH) [inline ctrl_pkg/tests/integrationtest_ruddercontroller.py:58]: What if the callback is not fast enough? Will it fail?

File: ctrl_pkg/tests/libtest_sail.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        self.assertAlmostEqual(calculate_sail_angle(0, max_roll, sail_limits), 0)
+
+    def test_positive_angle(self):
+        sail_limits = math.pi/1.4
+        max_roll = math.pi / 6
+        self.assertAlmostEqual(calculate_sail_angle(max_roll, max_roll, sail_limits), sail_limits)
+        self.assertEqual(calculate_sail_angle(max_roll/2, max_roll, sail_limits), sail_limits/2)
+        self.assertAlmostEqual(calculate_sail_angle(0, max_roll, sail_limits), 0)
+
+    def sail_trimming_test(self):
```
Comment (AutoSail-MDH) [inline ctrl_pkg/tests/libtest_sail.py]: This test is not run since it misses the prefix test_...

**Review by AutoSail-MDH — CHANGES_REQUESTED** [review]

**Comment by AutoSail-MDH** [comment]
Remove the unneccessary files such as __pycache__ and the stdmsg.msg

File: ctrl_pkg/tests/integrationtest_ruddercontroller.py (hunk 6 above)
Comment (jnn16012) [reply ctrl_pkg/tests/integrationtest_ruddercontroller.py:58]: No, the test only checks if the rudder turns to the stated position

File: ctrl_pkg/src/ctrl/pid.py (hunk 5 above)
Comment (jnn16012) [reply ctrl_pkg/src/ctrl/pid.py:89]: Fixed

**Pull request closed without merging**